sparse_ho/criterion.py

```diff
@@ -20,178 +20,157 @@ class HeldOutMSE():
     # cv as [(train, test)] ie directly the indices of the train
     # and test splits.
 
-    def __init__(self, X_val, y_val, model, convexify=False,
-                 gamma_convex=1e-2, X_test=None, y_test=None):
+    def __init__(
+            self, idx_train, idx_val, X_test=None, y_test=None,):
         """
         Parameters
         ----------
-        X_val : {ndarray, sparse matrix} of shape (n_samples, n_features)
-            Validation data
-        y_val : {ndarray, sparse matrix} of shape (n_samples,)
-            Validation target
-        model: object of the class Model (e.g. Lasso or SparseLogreg)
+        idx_train: np.array
+            indices of the training set
+        idx_test: np.array
+            indices of the testing set
         X_test : {ndarray, sparse matrix} of shape (n_samples_test, n_features)
             Test data
-        convexify: bool
-            this param should be remove from here XXX
-        gamma_convex: bool
-            this param should be removed from here XXX
         y_test : {ndarray, sparse matrix} of (n_samples_test)
             Test target
         """
-        self.X_val = X_val
-        self.y_val = y_val
         self.X_test = X_test
         self.y_test = y_test
-        self.model = model
-        self.convexify = convexify
-        self.gamma_convex = gamma_convex
+        self.idx_train = idx_train
+        self.idx_val = idx_val
 
         self.mask0 = None
         self.dense0 = None
         self.quantity_to_warm_start = None
         self.val_test = None
         self.rmse = None
 
-    def get_v(self, mask, dense):
-        return 2 * (self.X_val[:, mask].T @ (
-            self.X_val[:, mask] @ dense - self.y_val)) / self.X_val.shape[0]
-
-    def value(self, mask, dense):
-        val = (
-            norm(self.y_val - self.X_val[:, mask] @ dense) ** 2 /
-            self.X_val.shape[0])
-        return val
-
-    def value_test(self, mask, dense):
-        if self.X_test is not None and self.y_test is not None:
-            self.val_test = (
-                norm(self.y_test - self.X_test[:, mask] @ dense) ** 2 /
-                self.X_test.shape[0])
-        else:
-            self.val_test = None
-
-    def compute_rmse(self, mask, dense, beta_star):
-        if beta_star is not None:
-            diff_beta = beta_star.copy()
-            diff_beta[mask] -= dense
-            self.rmse = norm(diff_beta)
-        else:
-            self.rmse = None
+    def value_outer_crit(self, X, y, mask, dense):
+        """Compute the MSE on the validation set.
+        """
+        if X is None or y is None:
+            return None
+        return norm(y - X[:, mask] @ dense) ** 2 / len(self.idx_val)
```

Comment:
len(y) would be safer

so t this function, you pass full X_test and y_test, not full X and y, right ?

Reply:
> so t this function, you pass full X_test and y_test, not full X and y, right ?

yes